File: src/jobdu1545.java
import java.util.ArrayList;
import java.util.Comparator;
import java.util.PriorityQueue;
import java.util.Queue;
import java.util.Scanner;

public class jobdu1545 {
	public static class Edge {
		int from;
		int to;
		int v;

		public Edge(int f, int t, int v) {
			this.from = f;
			this.to = t;
			this.v = v;
		}
	}

	public static class Node {
		ArrayList<Edge> adjs = new ArrayList<Edge>();
		int minMax = Integer.MAX_VALUE;
	}

	public static class Step {
		int node;
		int m;

		public Step(int node, int m) {
			this.node = node;
			this.m = m;
		}
	}

	static int n = 0;
	static final int M = 10005;
	static Node[] nodes = null;
	static Edge[] edges = null;
	static Queue<Step> q = new PriorityQueue<Step>(10, new Comparator<Step>() {
		public int compare(Step o1, Step o2) {
			return o1.m - o2.m;
		}
	});

	static int findMaxMinConnect() {
		q.clear();
		int lastMax = Integer.MIN_VALUE;
		int min = Integer.MAX_VALUE;
		q.offer(new Step(0, Integer.MIN_VALUE));
		while (!q.isEmpty()) {
			Step s = q.poll();
			int c = s.node;
			lastMax = s.m;
			Node node = nodes[c];
			if (c == n - 1) {
				min = Math.min(min, lastMax);
				continue;
			}
			ArrayList<Edge> adjs = node.adjs;
			for (Edge e : adjs) {
				int m = Math.max(lastMax, e.v);
				if (m < nodes[e.to].minMax) {
					nodes[e.to].minMax = m;
					q.offer(new Step(e.to, m));
				}
			}
		}
		return min;
	}

	public static void main(String[] args) {
		int m;
		Scanner sc = new Scanner(System.in);
		while (sc.hasNextInt()) {
			n = sc.nextInt();
			m = sc.nextInt();
			nodes = new Node[n];
			for (int i = 0; i < n; ++i) {
				nodes[i] = new Node();
				nodes[i].minMax = Integer.MAX_VALUE;
			}
			edges = new Edge[m * 2];
			for (int i = 0; i < m; ++i) {
				int from = sc.nextInt();
				int to = sc.nextInt();
				int v = sc.nextInt();
				--from;
				--to;
				edges[2 * i] = new Edge(from, to, v);
				nodes[from].adjs.add(edges[2 * i]);

				edges[2 * i + 1] = new Edge(to, from, v);
				nodes[to].adjs.add(edges[2 * i + 1]);
			}
			int ret = findMaxMinConnect();
			if (ret == Integer.MAX_VALUE) {
				System.out.println(-1);
			} else {
				System.out.println(ret);
			}
		}
	}
}

File: cplusplus/jobdu1545.cpp
#include <cstdio>
#include <cstdlib>
#include <vector>
#include <iostream>
#include <queue>
#include <cstring>
#include <algorithm>
#include <limits>
#include <cmath>
using namespace std;
#define INT_MAX (1<<30)
#define INT_MIN (-(1<<30))
class Edge {
public:
	Edge() {
	}
	int v;
	int to;
};
class Node {
public:
	vector<int> adjs;
	int minMax;
};
struct Step {
	int node;
	int maxEdge;
};
const int M = 100005;
const int N = 10005;
int n, m;
Edge edges[2 * M];
Node nodes[N];
struct compare {
	bool operator()(const Step& l, const Step& r) {
		return l.maxEdge > r.maxEdge;
	}
};
int findmin() {
	priority_queue<Step, deque<Step> , compare> q;
	Step s;
	s.node = 0;
	s.maxEdge = INT_MIN;
	int min = INT_MAX;
	q.push(s);
	while (!q.empty()) {
		s = q.top();
		q.pop();
		int t = s.node;

		cout << "debug:" << nodes[t].minMax << endl;
		int currentMax = s.maxEdge;
		if (t == n - 1 && min > nodes[t].minMax) {
			min = nodes[t].minMax;
			continue;
		}
		vector<int>& adjs = nodes[t].adjs;
		vector<int>::iterator it = adjs.begin();
		while (it != adjs.end()) {
			int e = *it++;
			Edge &edge = edges[e];
			int maxEdge = max(currentMax, edge.v);
			if (nodes[edge.to].minMax > maxEdge) {
				nodes[edge.to].minMax = maxEdge;
				Step newStep;
				newStep.node = edge.to;
				newStep.maxEdge = maxEdge;
				q.push(newStep);
			}
		}
	}
	return min == INT_MAX ? -1 : min;
}
/*
 int main() {
 while (scanf("%d %d", &n, &m) != EOF) {
 for (int i = 0; i < n; ++i) {
 nodes[i].adjs.clear();
 nodes[i].minMax = INT_MAX;
 }
 int maxV = INT_MIN;
 int minV = INT_MAX;
 for (int i = 0; i < m; ++i) {
 int from, to, v;
 scanf("%d %d %d", &from, &to, &v);
 --from;
 --to;
 edges[i << 1].v = v;
 edges[(i << 1) + 1].v = v;
 edges[i << 1].to = to;
 edges[(i << 1) + 1].to = from;
 nodes[from].adjs.push_back(i << 1);
 nodes[to].adjs.push_back((i << 1) + 1);
 if (maxV < v)
 maxV = v;
 if (minV > v)
 minV = v;
 }
 cout << findmin() << endl;
 }
 return 0;
 }*/
